src/webportal/src/app/components/monaco-panel.jsx

```diff
@@ -0,0 +1,163 @@
+  // monaco variables
+  const monaco = useRef(null);
+  const editor = useRef(null);
+  const completionList = useRef({
+    suggestions: [],
+  });
+
+  // resize event
+  const handleResize = () => {
+    if (editor.current !== null) {
+      editor.current.layout();
+    }
+  };
+  useEffect(() => {
+    window.addEventListener('resize', handleResize);
+    return () => {
+      window.removeEventListener('resize', handleResize);
+    };
+  });
+
+  // json schema
+  const setSchemas = (monaco, schemas) => {
+    if (schemas !== null) {
+      monaco.languages.json.jsonDefaults.setDiagnosticsOptions({
+        validate: true,
+        schemas,
+      });
+    } else {
+      monaco.languages.json.jsonDefaults.setDiagnosticsOptions({
+        validate: false,
+        schemas: [],
+      });
+    }
+  };
+  useLayoutEffect(() => {
+    if (monaco.current !== null) {
+      setSchemas(monaco.current, schemas);
+    }
+  }, [schemas]);
+
+  // completion items
+  useEffect(() => {
+    if (completionItems) {
+      completionList.current = {
+        suggestions: completionItems.map((x) => ({
+          label: x,
+          insertText: x,
+        })),
+      };
+    }
+  }, [completionItems]);
+
+  return (
+    <div>
+      <Panel
+        onDismiss={onDismiss}
+        isLightDismiss={true}
+        isOpen={isOpen}
+        type={PanelType.large}
+        headerText={title}
+        styles={{
+          main: [ColorClassNames.neutralPrimaryBackground],
+          headerText: [ColorClassNames.white],
+          overlay: [ColorClassNames.blackTranslucent40Background],
+          content: [t.flex, t.flexAuto],
+          scrollableContent: [t.flex, t.flexAuto, {overflowY: 'visible'}],
+          closeButton: [ColorClassNames.white, ColorClassNames.neutralQuaternaryHover],
+        }}
+      >
+        {header && <div className={c(t.mb4, t.flex)}>
+          {header}
+        </div>}
+        <div className={c(t.flexAuto, t.flex, t.flexColumn)}>
+          <div className={c(monacoHack)} style={{flex: '1 1 100%', minHeight: 0}}>
+            <MonacoEditor
+              className={c(t.flexAuto)}
+              theme='vs-dark'
+              language='text'
+              options={{
+                wordWrap: 'on',
+                readOnly: true,
+              }}
+              editorDidMount={(e, m) => {
+                // save monaco context
+                editor.current = e;
+                monaco.current = m;
+                // completion provider
+                for (const lang of ['json', 'yaml', 'plaintext']) {
+                  monaco.current.languages.registerCompletionItemProvider(lang, {
+                    provideCompletionItems() {
+                      return completionList.current;
+                    },
+                  });
+                }
+                // json schema
+                setSchemas(monaco.current, schemas);
```

Comment:
I'm not sure how to use this schema. I only can use it when language is Json? If I set the language to yaml, how can I validate the content? By customize the onDismiss function?

Reply:
MonacoEditor's built in schema check only support `language=json`.

If you want to use your own validation logic (yaml/joi)
-  use [react-monaco-editor](https://github.com/react-monaco-editor/react-monaco-editor)'s `onChange` event to get the content.
    - use `monacoProps` to pass your callback func
-  display the validation result in header area.